web/modules/custom/search_api_field/src/Plugin/facets/facet_source/SearchApiField.php

```diff
@@ -0,0 +1,153 @@
+    parent::__construct($configuration, $plugin_id, $plugin_definition, $query_type_plugin_manager, $search_results_cache);
+
+    // Load facet plugin definition and depending on those settings; load the
+    // corresponding search api page and load its index.
+    $field_id = $plugin_definition['search_api_field'];
+    /* @var $page \Drupal\search_api\Utility\QueryHelper */
+    $field = FieldStorageConfig::load($field_id);
+    $index = $field->getSetting('index');
+    $this->index = Index::load($index);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function fillFacetsWithResults($facets) {
+    // Check if there are results in the static cache.
+    $results = $this->searchApiQueryHelper->getResults($this->pluginId);
+
+    // If there are no results, execute the search page and check for results
+    // again. This happens when a page or block is cached, so Search API has
+    // not fired an actual search.
+    if (!$results) {
+      /* @var $search_api_index \Drupal\search_api\IndexInterface */
+      $search_api_index = $this->getIndex();
+
+      // Create the query.
+      $options = [
+        'parse_mode' => 'direct',
+        // @Todo Fix limit, get it from field settings.
+        'limit' => 10,
+        'offset' => isset($_GET['page']) ? $_GET['page'] : 0,
+        'search id' => $this->pluginId,
+      ];
+      $query = $search_api_index->query($options);
+
+      // Keys.
+      $keys = \Drupal::request()->get('keys');
+      if (!empty($keys)) {
+        $query->keys($keys);
+      }
+
+      // Index fields.
+      $query->setFulltextFields();
+
+      // Execute the query.
+      $results = $query->execute();
+    }
+
+    // If we got results from the cache, this is the first code executed in this
+    // method, so it's good to double check that we can actually work with
+    // $results.
+    if ($results instanceof ResultSetInterface) {
+      // Get our facet data from the results.
+      $facet_results = $results->getExtraData('search_api_facets');
+
+      // Loop over each facet and execute the build method from the given query
+      // type.
+      foreach ($facets as $facet) {
+        if (isset($facet_results[$facet->getFieldIdentifier()])) {
+          $configuration = array(
+            'query' => NULL,
+            'facet' => $facet,
+            'results' => $facet_results[$facet->getFieldIdentifier()],
+          );
+
+          // Get the Facet Specific Query Type so we can process the results
+          // using the build() function of the query type.
+          $query_type = $this->queryTypePluginManager->createInstance($facet->getQueryType(), $configuration);
+          $query_type->build();
+        }
+      }
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getPath() {
+    // @todo Is this always right?
+    $path = \Drupal::service('path.current')->getPath();
```

Comment:
But you can inject the `CurrentPathStack` instead of leeching it from the global scope!

I am not familiar with this service, but it has this big warning:

```
/**
 * Represents the current path for the current request.
 *
 * Note: You should not rely on paths but rather on route names / parameters or
 *   other indicators like context. For some fundamental parts, like routing or
 *   path processing, there is unfortunately no way around dealing with paths.
 */
```

Maybe you can get it from the `RequestStack`? I don't know this by heart though so you'll need to do some digging. But if this works for now, fine ;)